### PR TITLE
Implements `blobs_last_used_by` for grouped exchanges.

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

There are some `blob_last_used_by` operations that can be grouped. 

## Proposal

The following changes were done:
* The `blobs_last_used_by` is introduced to the `Validator` trait and to each type implementing it.

The use in the protocol is for a subsequent PR with the target being updater.rs. 

## Test Plan

No test so far for this feature.

## Release Plan

It could be used in the TestNet / DevNet.

## Links

None.

### CODE REVIEW

File: linera-core/src/unit_tests/test_utils.rs [hunk 1]
```diff
@@ -503,6 +513,27 @@ where
...
+            .storage_client()
+            .read_blob_states(&blob_ids)
+            .await;
+        let hashes = match blob_states {
+            Err(err) => Err(err.into()),
+            Ok(blob_states) => Ok(blob_states
+                .into_iter()
+                .map(|blob_state| blob_state.last_used_by)
+                .collect()),
+        };
```
Comment (andresilva91) [inline linera-core/src/unit_tests/test_utils.rs:534]: This seems cleaner/more idiomatic:

```suggestion
        let hashes = blob_states
            .map(|states| states.into_iter().map(|state| state.last_used_by).collect())
            .map_err(Into::into);
```

File: linera-rpc/proto/rpc.proto [hunk 2]
```diff
@@ -69,6 +69,9 @@ service ValidatorNode {
 
   // Returns the hash of the `Certificate` that last used a blob.
   rpc BlobLastUsedBy(BlobId) returns (CryptoHash);
+
+  // Returns the hash of the `Certificate` that last used a blob.
```
Comment (andresilva91) [inline linera-rpc/proto/rpc.proto:73]:
```suggestion
  // Returns the hashes of the certificates that last used each of the blobs.
```

File: linera-rpc/proto/rpc.proto [hunk 3]
```diff
@@ -272,6 +279,11 @@ message BlobId {
   bytes bytes = 1;
 }
 
+// A content-addressed blob ID i.e. the hash of the `BlobContent`.
```
Comment (andresilva91) [inline linera-rpc/proto/rpc.proto:282]:
```suggestion
// Content-addressed blob IDs i.e. the hashes of the `BlobContent`s.
```

File: linera-service/src/proxy/grpc.rs [hunk 4]
```diff
@@ -531,6 +531,25 @@ where
...
+        let blob_ids: Vec<linera_base::identifiers::BlobId> = request.into_inner().try_into()?;
+        let blob_states = self
+            .0
+            .storage
+            .read_blob_states(&blob_ids)
+            .await
+            .map_err(|err| Status::from_error(Box::new(err)))?;
+        let results = blob_states
+            .into_iter()
+            .map(|result| result.last_used_by)
```
Comment (andresilva91) [inline linera-service/src/proxy/grpc.rs:549]:
```suggestion
            .map(|blob_state| blob_state.last_used_by)
```

File: linera-service/src/proxy/main.rs [hunk 5]
```diff
@@ -320,6 +320,14 @@ where
             BlobLastUsedBy(blob_id) => Ok(Some(RpcMessage::BlobLastUsedByResponse(Box::new(
                 self.storage.read_blob_state(*blob_id).await?.last_used_by,
             )))),
+            BlobsLastUsedBy(blob_ids) => Ok(Some(RpcMessage::BlobsLastUsedByResponse(Box::new(
+                self.storage
+                    .read_blob_states(&blob_ids)
+                    .await?
+                    .into_iter()
+                    .map(|x| x.last_used_by)
```
Comment (andresilva91) [inline linera-service/src/proxy/main.rs:328]:
```suggestion
                    .map(|blob_state| blob_state.last_used_by)
```

File: linera-core/src/node.rs [hunk 6]
```diff
@@ -103,6 +103,10 @@ pub trait ValidatorNode {
 
     /// Returns the hash of the `Certificate` that last used a blob.
     async fn blob_last_used_by(&self, blob_id: BlobId) -> Result<CryptoHash, NodeError>;
+
+    /// Returns the hashes of the `Certificates` that last used some blobs.
```
Comment (andresilva91) [inline linera-core/src/node.rs:107]: AFAIK we use back quotes for actual type names
```suggestion
    /// Returns the hashes of the `Certificate`s that last used some blobs.
```

Comment (MathieuDutSik) [reply linera-core/src/node.rs:107]: Ok.

File: linera-service/src/proxy/main.rs (hunk 5 above)
Comment (MathieuDutSik) [reply linera-service/src/proxy/main.rs:328]: Ok.

File: linera-service/src/proxy/grpc.rs (hunk 4 above)
Comment (MathieuDutSik) [reply linera-service/src/proxy/grpc.rs:549]: Ok.

File: linera-rpc/proto/rpc.proto (hunk 3 above)
Comment (MathieuDutSik) [reply linera-rpc/proto/rpc.proto:282]: Ok.

File: linera-rpc/proto/rpc.proto (hunk 2 above)
Comment (MathieuDutSik) [reply linera-rpc/proto/rpc.proto:73]: Ok.

File: linera-core/src/unit_tests/test_utils.rs (hunk 1 above)
Comment (MathieuDutSik) [reply linera-core/src/unit_tests/test_utils.rs:534]: Correct. Done.

File: linera-rpc/src/message.rs [hunk 7]
```diff
@@ -30,6 +30,7 @@ pub enum RpcMessage {
     DownloadCertificate(Box<CryptoHash>),
     DownloadCertificates(Box<Vec<CryptoHash>>),
     BlobLastUsedBy(Box<BlobId>),
```
Comment (afck) [inline linera-rpc/src/message.rs:32]: Do we need the singular versions of these RPCs at all? (Maybe something to discuss for later.)

File: linera-rpc/src/message.rs [hunk 8]
```diff
@@ -190,6 +196,18 @@ impl TryFrom<RpcMessage> for Vec<Certificate> {
     }
 }
 
+impl TryFrom<RpcMessage> for Vec<CryptoHash> {
+    type Error = NodeError;
+    fn try_from(message: RpcMessage) -> Result<Self, Self::Error> {
+        use RpcMessage::*;
```
Comment (afck) [inline linera-rpc/src/message.rs:202]: I'd prefer to not import enum variants.
(As @ndr-ds pointed out recently, that can be dangerous for ones without fields. But even for the ones with fields, I'd say it's safer to fully qualify them, so we're safe even if the fields are removed later.)

**Review by afck — APPROVED** [review]

File: linera-rpc/src/message.rs (hunk 7 above)
Comment (MathieuDutSik) [reply linera-rpc/src/message.rs:32]: Maybe we can remove them in a subsequent PR.

File: linera-rpc/src/message.rs (hunk 8 above)
Comment (MathieuDutSik) [reply linera-rpc/src/message.rs:202]: So, we would like to have a type
```rust
struct CryptoHashes {
  hashes: Vec<CryptoHash>,
}
```
as well as similar for `Vec<BlobId>`.

So, that we so not have the implementation of `From/TryFrom` for `Vec<_>`?

Comment (afck) [reply linera-rpc/src/message.rs:202]: Not sure; that's not what I meant. I'd just remove the `use RpcMessage::*` and fully qualify the enum variant below.

Comment (MathieuDutSik) [reply linera-rpc/src/message.rs]: Ok, thank you. It is good that I asked about that problem.

**Pull request merged**